### PR TITLE
Device Config backup workflow module

### PR DESCRIPTION
**Pull request by sonalideepthi777**
## Description
Update the Device backup config workflow module to support for replacement of APIs, and improved functionality with version-based routing.

## Type of Change
- [ ] Bug fix
- [x] New feature
- [ ] Breaking change
- [x] Documentation update

## Checklist
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules
- [x] All the sanity checks have been completed and the sanity test cases have been executed

## Ansible Best Practices
- [x] Tasks are idempotent (can be run multiple times without changing state)
- [ ] Variables and secrets are handled securely (e.g., using `ansible-vault` or environment variables)
- [x] Playbooks are modular and reusable
- [x] Handlers are used for actions that need to run on change

## Documentation
- [x] All options and parameters are documented clearly.
- [x] Examples are provided and tested.
- [x] Notes and limitations are clearly stated.

### CODE REVIEW

File: plugins/modules/device_configs_backup_workflow_manager.py [hunk 1]
```diff
@@ -394,6 +398,11 @@ class Device_configs_backup(DnacBase):
     def __init__(self, module):
         super().__init__(module)
         self.skipped_devices_list = []
+        self.payload = module.params
+        self.keymap = {}
+        self.dnac_version = int(self.payload.get(
```
Comment (madhansansel) [inline plugins/modules/device_configs_backup_workflow_manager.py:403]: setting dnac version is already done in dnac.py.. Can you check?

https://github.com/madhansansel/dnacenter-ansible/blob/main/plugins/module_utils/dnac.py
        self.payload = module.params
        self.dnac_version = int(self.payload.get("dnac_version").replace(".", ""))
        # Dictionary to store multiple versions for easy maintenance and scalability
        # To add a new version, simply update the 'dnac_versions' dictionary with the new version string as the key
        # and the corresponding version number as the value.
        self.dnac_versions = {
            "2.3.5.3": 2353,
            "2.3.7.6": 2376,
            "2.2.3.3": 2233
            # Add new versions here, e.g., "2.4.0.0": 2400
        }

        # Dynamically create variables based on dictionary keys
        for version_key, version_value in self.dnac_versions.items():
            setattr(self, "version_" + version_key.replace(".", "_"), version_value)

File: plugins/modules/device_configs_backup_workflow_manager.py [hunk 2]
```diff
@@ -514,109 +523,101 @@ def validate_site_exists(self, site_name):
...
-            self.log("Response received post 'get_site' API call: {0}".format(str(response)), "DEBUG")
-            response = response.get("response")
-
-            # Process the response if available
-            if response:
-                site_id = response[0].get("id")
-                site_exists = True
-            else:
-                self.log("No response received from the 'get_site' API call.", "WARNING")
+            response = self.get_site(site_name)
```
Comment (madhansansel) [inline plugins/modules/device_configs_backup_workflow_manager.py:529]: We don't need this API at all. Check get_site_id API and use it.. if you need site_exists also, then change the get_site_id API

File: plugins/modules/device_configs_backup_workflow_manager.py [hunk 3]
```diff
@@ -514,109 +523,101 @@ def validate_site_exists(self, site_name):
...
-            self.update_result("failed", False, self.msg, "ERROR")
+            self.status = "failed"
+            self.msg = ("An exception occurred: Site '{0}' does not exist in the Cisco Catalyst Center.".format(site_name))
+            self.result['response'] = self.msg
+            self.log(self.msg, "ERROR")
             self.check_return_status()
 
         return (site_exists, site_id)
 
     def get_device_ids_from_site(self, site_name, site_id):
```
Comment (madhansansel) [inline plugins/modules/device_configs_backup_workflow_manager.py:545]: Move this API to dnac.py and make it common....

File: plugins/modules/device_configs_backup_workflow_manager.py [hunk 4]
```diff
@@ -514,109 +523,101 @@ def validate_site_exists(self, site_name):
...
-            # Log an error message if any exception occurs during the process
-            self.log("Unable to fetch the device(s) associated to the site '{0}' due to {1}".format(site_name, str(e)), "ERROR")
+            self.log("Unable to fetch the device(s) associated with the site '{0}' due to {1}".format(site_name, str(e)), "ERROR")
 
-        # Log a warning if no reachable devices are found
         if not mgmt_ip_to_instance_id_map:
-            self.log("Reachable devices not found at Site: {0}".format(site_name), "WARNING")
+            self.msg = "No reachable devices found at Site: {0}".format(site_name)
+            self.update_result("ok", False, self.msg, "INFO")
+            self.module.exit_json(**self.result)
```
Comment (madhansansel) [inline plugins/modules/device_configs_backup_workflow_manager.py:620]: Why do we need to exit from the module if devices not found at site?

File: plugins/modules/device_configs_backup_workflow_manager.py (hunk 1 above)
Comment (sonalideepthi777) [reply plugins/modules/device_configs_backup_workflow_manager.py]: Removed

File: plugins/modules/device_configs_backup_workflow_manager.py (hunk 2 above)
Comment (sonalideepthi777) [reply plugins/modules/device_configs_backup_workflow_manager.py]: Removed that API and directly using get_site_id

File: plugins/modules/device_configs_backup_workflow_manager.py (hunk 3 above)
Comment (sonalideepthi777) [reply plugins/modules/device_configs_backup_workflow_manager.py]: Moved, I added two functions in dnac.py: one for getting device IDs from a site and another for retrieving the IP address from a device id.

File: plugins/modules/device_configs_backup_workflow_manager.py (hunk 4 above)
Comment (sonalideepthi777) [reply plugins/modules/device_configs_backup_workflow_manager.py]: If the device ID is not available, then the IP address cannot be retrieved in this task

File: plugins/modules/device_configs_backup_workflow_manager.py (hunk 2 above)
Comment (madhansansel) [reply plugins/modules/device_configs_backup_workflow_manager.py:524]: We can directly use get_site_id rite? 
Why do we use get_site first, and use get the site id from the response?

Comment (sonalideepthi777) [reply plugins/modules/device_configs_backup_workflow_manager.py:524]: I removed and using get_site_id directly

Comment (madhansansel) [reply plugins/modules/device_configs_backup_workflow_manager.py:524]: Can you update the code?

Comment (madhansansel) [reply plugins/modules/device_configs_backup_workflow_manager.py:524]: We still don't need this API. What is the purpose of the API?

Comment (sonalideepthi777) [reply plugins/modules/device_configs_backup_workflow_manager.py]: Updated

**Review by madhansansel — APPROVED** [review]

**Pull request merged**